### ISSUE TITLE
Difference between RFECV and SFS ?

### ISSUE THREAD

**Issue by lschneidpro** [rasbt/mlxtend#582]
Hi there,

I used scikit.REFCV and mlxtend.SFS (backward) on the same data, same classifier, same cv, same scorer,... I also did a third version with sample weights passed to SFS's estimator

![RFE](https://user-images.githubusercontent.com/40396043/63971921-f0cfa500-caa7-11e9-8ba0-0075d81c19dd.png)

And i'm conflicted to see that many discrepencies. Should the largest subsets  have at least the more or less same CV score ?

From my unterstanding, RFECV and SFS backward should work the same except the first one selects subset based on feature importance and the latter one on the metric provided.

Thanks for any input.

**Comment by rasbt** [comment]
If there is no or  not a lot of interaction between the different features, and if a linear model can get good predictive performance on a given dataset, then SFS and RFE can indeed yield similar results. However, both approaches are fundamentally different. RFE selects based on linear model coefficients and SFS selects based on model performance. Dependent on the dataset these can but don’t have to be leading to the same results. Also, RFE will become useless if you have no linear problems, think of XOR

**Comment by lschneidpro** [comment]
> If there is no or not a lot of interaction between the different features, and if a linear model can get good predictive performance on a given dataset, then SFS and RFE can indeed yield similar results. However, both approaches are fundamentally different. RFE selects based on linear model coefficients and SFS selects based on model performance. Dependent on the dataset these can but don’t have to be leading to the same results. Also, RFE will become useless if you have no linear problems, think of XOR

I totally understand it but:

- They should at least yield the same cv score when all features are included at the begining of both processes (see the 4 plots on the left)
- How can I interpret that RFECV is smooth and SFS is very noisy?

By the way, I'm using lightgbm's GBT. So no linear coefficients, but feature importance

Thank you

**Comment by rasbt** [comment]
These are all good questions but hard to answer based on the plots. 

> They should at least yield the same cv score when all features are included at the begining of both processes (see the 4 plots on the left)

It depends on how you compute the cv score. I assume you used the same number of folds, but what kind of models did you compare? Did you use the same model in both cases? If so, you can then also feed in the model + the dataset with all features and see which one gives the same results compared to calling `cross_val_score`, which should provide the "correct" score in any case. (There could be some variation based on random seeds for splitting into folds).

> How can I interpret that RFECV is smooth and SFS is very noisy?

What models/ml algorithms are you using? Maybe there's some overfitting.

**Comment by lschneidpro** [comment]
> It depends on how you compute the cv score. I assume you used the same number of folds, but what kind of models did you compare? Did you use the same model in both cases? If so, you can then also feed in the model + the dataset with all features and see which one gives the same results compared to calling `cross_val_score`, which should provide the "correct" score in any case. (There could be some variation based on random seeds for splitting into folds).

I used my own cv generator (to deal with time dependent data). I can maybe implement a fixed random seed.

> What models/ml algorithms are you using? Maybe there's some overfitting.
I'm using lightGBM's classifier (type GBT), maybe i should go for a RF. But How would it be possible to implement nested-CV to do hyper-tuning to reduce overfitting ?

Thanks

**Comment by rasbt** [comment]
> I used my own cv generator (to deal with time dependent data). I can maybe implement a fixed random seed.

That might explain the difference. For testing/comparison, if you think there is a bug somewhere, it would be best to do all experiments with a fixed random seed.

> I'm using lightGBM's classifier (type GBT), maybe i should go for a RF. But How would it be possible to implement nested-CV to do hyper-tuning to reduce overfitting ?

Are you referring specifically to the spikes in the plot you provided? How are these scores computed? Are these internal scores (center column) computed via the model inside the SFS, or are these scores you computed after SFS based on the selected features?

**Comment by lschneidpro** [comment]
> That might explain the difference. For testing/comparison, if you think there is a bug somewhere, it would be best to do all experiments with a fixed random seed.

Yes, I just implemented. I'm waiting to have all information about what is going one, before running SFS. Running SFS on the largest subset took me easily 6h+ on a 48 cores instance.

> Are you referring specifically to the spikes in the plot you provided? How are these scores computed? Are these internal scores (center column) computed via the model inside the SFS, or are these scores you computed after SFS based on the selected features?

Yes they are both the CV scores stored by the respective objects.

I have still a validation and test set aside. My plan was to take different promising subsets from RFECV and SFS and retrain them with HP tuning. Finally, I'd check the scores on my validation set, the test set being the set to evaluate the score of my entire machine learning pipeline.

Few remarks:

1. I'm aware that might be some selection bias in this process. But let's remain pragmatic. Its computationally infeasible to do pre-processing, model selection, HP tuning, feature selection,... on each single loop of CV (See https://www.nodalpoint.com/not-perform-feature-selection/). In addition, the main idea was still do interpret the feature selection (It is for a thesis).  Using pipelines, I'd just retrieve the best pipeline and not a full study case of the feature selection

2. About the overfitting part, I believe RFECV and SFS have been initially designed to mainly find the best subset and pass it to the next steps in a pipeline. All the examples provided show RFE  for models with inital parameters. As the subset keep changing, there is a clear overfitting as we could expect different optimal HP for each subsets and thus CV score would be ideally different. 
I asked about the possibility to do nested CV for HP alongside RFS, to be sure my SFS CV score cannot be suspected of overfitting.
From http://www.feat.engineering/recursive-feature-elimination.html, it is said that for feature selection good practices: 
  > For models that are sensitive to the choice of tuning parameters, there may be a need to  retune the model when each subset is evaluated. In this case, a separate, nested resampling  process is required to tune the model. In many cases, the computational costs may make the search tools practically infeasible.
As such, the authors performed RFE on RF without nested tuning. I assumed that RFE with GBT could be done in the same way.


So finally what should I do ? :)

Thank you

**Comment by rasbt** [comment]
Regarding the suspected bug (discrepancy of measured performance on same feature subsets when SFS and RFECV is used), I would investigate it as follows:

SFS

- Make a smaller subset of your dataset so that it doesn't take too long to run
- Try the backward selection, and then maybe just select m-1 features (m=number of features in your dataset) 
- Look at the performance of the initial subset m and the selected subset m-1.

RFECV

- Do the same thing as above. See if the initial subset performance matches the subset performance in SFS. 
- if the subsets from SFS and RFECV at m-1 features are the same, also compare these


If there is a discrepancy, run the previous steps 

- Try the backward selection, and then maybe just select m-1 features (m=number of features in your dataset) 
- Look at the performance of the initial subset m and the selected subset m-1.

multiple times. Maybe it's that your CV makes different splits (due to random seeds), which is why you might observe differences.


--- 

In practice, the performance in

![Screen Shot 2019-08-30 at 11 37 40 AM](https://user-images.githubusercontent.com/5618407/64037167-958aca80-cb1a-11e9-8d71-ba296c463414.png)

doesn't look too bad. Except the weirdness at ~75 and ~100, it basically tells you that there is no benefit in selecting more than ~10 features as the performance doesn't really improve after that.

**Comment by lschneidpro** [comment]
> Here, I am assuming that one of the lines is the test fold performance within the SFS. I am not sure though what you mean/did with the correlation filter etc.

The blue is all my features.
The orange is the same features, but I removed some to not have correlation above 0.8
The green is also the same starting features, but I filtered for dependency (See https://explained.ai/rf-importance/index.html#corr_collinear)

The motive for doing RFE/SFS on three different intial subsets was to spot feature interaction, as we can clearly see it from RFECV plots. Removing correlated features decreased performances, so there was some benefical correlation.

**Comment by rasbt** [comment]
Ok, I see, so the curves are just from different initial starting sets. Makes sense :).


Coming back to the part where you said that the plots are based on the internal scores. What confuses me though is the following statement then:

> I'm using lightGBM's classifier (type GBT), maybe i should go for a RF. But How would it be possible to implement nested-CV to do hyper-tuning to reduce overfitting ?

How can you run RFECV with lightGBM or RF, these are nonparametric models, right?! So, they don't have model coefficients. Afaik the RFECV, similar to the RFE, needs a parametric model, and in particular, an `Estimator` with a `coef_ ` attribute, because it selects the features based on the model weights.

**Comment by lschneidpro** [comment]
> How can you run RFECV with lightGBM or RF, these are nonparametric models, right?! So, they don't have model coefficients. Afaik the RFECV, similar to the RFE, needs a parametric model, and in particular, an `Estimator` with a `coef_ ` attribute, because it selects the features based on the model weights.

They included possibility for models with feature_importances (https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFECV.html):

> **estimator** : object
> A supervised learning estimator with a fit method that provides information about feature importance either through a coef_ attribute or through a feature_importances_ attribute.

**Comment by rasbt** [comment]
> They included possibility for models with feature_importances

nice, didn't know that! If it doesn't take too long, would it be possible to make these plots with a simple logistic regression model?

**Comment by lschneidpro** [comment]
on both SFE/RFECV? what would be the motive ?

**Comment by rasbt** [comment]
I am curious regarding the fluctuation in the SFS

**Comment by lschneidpro** [comment]
OK RFE/SFS running with LR

**Comment by lschneidpro** [comment]
![LR](https://user-images.githubusercontent.com/40396043/64261289-93ff3f00-cf2c-11e9-92ac-b3653acc9895.png)

**Comment by rasbt** [comment]
Oh, these are really nice plots. Also, like expected, the performance first improves by adding some new features ... then, at some point if too many features are included, it declines due to overfitting.

The fluctuation within lightgbm's GBT could be due to the several reasons, but does it  have a random seed that can be set to a fixed value?

**Comment by lschneidpro** [comment]
The algorithm or the CV itself?

**Comment by rasbt** [comment]
For the algo. but if you are using your own CV splitter, then ideally both for the CV and the algorithm

**Comment by lschneidpro** [comment]
> For the algo. but if you are using your own CV splitter, then ideally both for the CV and the algorithm

Is that really necessary? I have already 4 instance running it with random_state for CV but not for LGM :)

**Comment by rasbt** [comment]
not really necessary, but it may be helpful to find out why there are the fluctuations when doing the feature selection with the gbm

**Comment by lschneidpro** [comment]
ok I restarted

**Comment by lschneidpro** [comment]
There is still one curve computing, but i think we got the point

![GBT](https://user-images.githubusercontent.com/40396043/64330731-c0b96200-cfd1-11e9-9cab-fd83c4680d46.png)

**Comment by rasbt** [comment]
That looks pretty good to me. You basically get a ~0.975 ROC, and I don't think it can get much better than that :P.  So in addition to the feature re-scaling, you also fixed the random seed in the GBM?

**Comment by lschneidpro** [comment]
Actually the previous plot was wrong. Here it is.

Yes, Random_seed is fixed for CV and GBT

![GBT](https://user-images.githubusercontent.com/40396043/64485573-19336e00-d223-11e9-8d05-cd37573d7b27.png)

**Comment by rasbt** [comment]
Ok, that looks reasonable though. Again, the difference between the very first plot and this is the fixed random seed? Could be that the bigger fluctuation in the very first plot could then be just due to high variance.

**Comment by lschneidpro** [comment]
Yes the differences are

- fixed seed for the model and cv

- rolling scaling of samples along features ( to keep a certain homogeneity)
